src/instructions/mod.rs

```diff
@@ -0,0 +1,12 @@
+mod add;
+
+pub use add::Add;
+
+use crate::{memory::Memory, registers::Registers};
+
+pub trait Instruction {
+    fn from_bits(bits: u16) -> Self
+    where
+        Self: Sized;
+    fn execute(&self, registers: &mut Registers, memory: &mut Memory);
+}
```

Comment:
I believe you could rewrite this as:
```suggestion
pub trait Instruction: Sized {
    fn from_bits(bits: u16) -> Self;
    fn execute(&self, registers: &mut Registers, memory: &mut Memory);
}
```

Reply:
Regardless, I'm not sure it's necessary or useful for this to be a `trait`. You won't be able to instantiate different instructions, which may vary in size, using `from_bits` and store them to the same place and work abstractly, so you'll end up with a `match` either way. So, maybe this could be a `struct Instruction` or `enum Instruction` with an `impl Instruction`, which then can be built with `from_bits` and `execute`d.

Reply:
I agree, you could have a struct `Instruction` with an `Opcode` attribute and the functions it already has (`from_bits` and `execute`). In this case you would need to have 2 matching patterns (one per function). To prevent that -if you want to- an idea would be to have an enum `Instruction` with each opcode variant containing its particular executing function (i.e. `ADD(add)` with `add` being a `fn`). Also, [this](https://github.com/lazcanoluca/lc3-vm/pull/1#discussion_r1708158051) could work too.

Reply:
Thanks for the input! I was actually not fully pleased with my solution involving the `trait`. I went with the `enum` instead. The `enum` will be in charge of the `Instruction` abstraction. Let me know what you think!

Reply:
Looks better to me now!